### ISSUE TITLE
[P0] V4: add lossless mxfp4-rans256-g0 expert storage

### ISSUE THREAD

**Issue by mateocabanal** [mateocabanal/colibri#6]
## Goal
Reduce routed-expert disk bytes without changing model math by adding a byte-exact entropy-coded storage encoding around canonical MXFP4/fmt7.

## Why this is P0
Routed expert bytes cost both checkpoint storage and cache-miss bandwidth. Lossless compression can reduce both while preserving the existing matmul kernels and model quality exactly.

## Format direction
Add `mxfp4-rans256-g0` as a storage/container encoding, not a new mathematical quant format.

Encode separately:
- E2M1 weight nibbles
- E8M0 scale bytes

Reuse the existing rANS machinery and preserve compatibility of the current nibble codec. Add a sibling/generalized u8 path for scale bytes rather than nibble-splitting scales.

Initial design target:
```text
32-bit rANS state
RANS_L = 1<<23
scale_bits = 14
M = 16384
256 independent streams
static checkpoint/shard tables
```

## Phase 0 — measure V4 entropy first
Create a corpus-wide histogram tool over real V4 experts, split by:
- layer
- expert
- w1/w2/w3
- E2M1 nibbles
- E8M0 scales

Report actual achievable encoded ratio before freezing expectations. Earlier planning suggests ~0.70–0.82x combined bytes may be plausible, but the real V4 corpus is authoritative.

## Loader integration
Target flow:
```text
SSD compressed record
  -> decode exact E2M1/E8M0
  -> existing resident/rows16 preparation
  -> unchanged ColiExpertView
  -> unchanged matmul
```

LRU should store decoded experts, not compressed blobs. Loader lanes own reusable decode scratch; no per-miss record-sized malloc.

Later optimization may decode directly into final rows16/resident layout, but retain a canonical reference decode path for byte comparison.

## Metadata/format rules
- Explicit `colibri.fmt` metadata; compressed size is data-dependent.
- Do not infer record bytes from a fixed compression ratio.
- Existing fmt7 remains supported.
- No public quant ordinal for the lossless storage wrapper.
- Preserve required rANS readable slack/alignment contracts.

## Correctness gate
This is lossless. Therefore:
- [ ] decoded weight bytes are byte-identical to original MXFP4
- [ ] decoded E8M0 scale bytes are byte-identical
- [ ] rows16 output matches existing conversion
- [ ] routes/logits/tokens are identical
- [ ] corruption/truncation is detected safely

Any numerical difference is a bug, not an acceptable tolerance.

## Performance gate
Using #1, report:
- compressed/raw expert bytes
- rANS decode ms/miss
- exposed I/O wait ms/token
- cold/miss-heavy decode tok/s
- warm decode regression, if any

The codec succeeds when SSD-byte savings exceed decode overhead on realistic miss-heavy workloads.

## Follow-up
Only after this issue establishes the lossless baseline should lossy 3-bit experiments compare themselves against compressed MXFP4 rather than raw 4.25-bit MXFP4.

**Comment by mateocabanal** [comment]
### CSF integration update

The new Colibri Serving Format plan is tracked in #22–#28. This issue remains the authority for the **lossless MXFP4 rANS codec**, but it should no longer invent its own long-term container framing.

Integration target:
```text
#6 defines/implements codec semantics
        ↓
#22 supplies CSF math/scale/codec/layout fields
        ↓
#24 compiler writes independently compressed expert records
        ↓
#25 runtime decodes them into the expert LRU
        ↓
#26 may later decode directly into a native resident layout
```

Important boundary: keep rANS codec tests usable independently of CSF. The codec should accept explicit byte streams/metadata; CSF is just one storage container around it.

**Comment by mateocabanal** [comment]
### Compression-layout risk: 256 streams can trade I/O bytes for decode/setup overhead

The 256-way rANS direction is worth testing, but I would keep the stream count/table granularity empirical. For independently loadable expert records, many streams can add framing/table state, scatter the decoder's working set, and make direct decode into final target layout harder. On a fast SSD/UMA path the codec can become CPU-bound before the saved bytes translate into lower exposed stall.

Two things I would measure in Phase 0 in addition to entropy ratio:
1. encoded ratio **including all per-record/per-stream framing**;
2. decode GB/s and p95 miss latency for several stream counts (e.g. 16/32/64/128/256), not just the chosen 256.

Also watch context dependence: E2M1 symbol distributions may differ strongly by matrix/position/block, so one static table can leave compression on the table while too many local tables erase the gain in metadata/cache misses. The best serving point may be fewer streams/tables than the entropy-only optimum.
